client/network/src/protocol/sync.rs

```diff
@@ -687,7 +718,21 @@ impl<B: BlockT> ChainSync<B> {
 				return None
 			}
 
-			if let Some((range, req)) = peer_block_request(
+			// If our best queued is more than `MAX_BLOCKS_TO_LOOK_BACKWARDS` blocks away from the
+			// common number, the peer best number is higher than our best queued and the common
+			// number is smaller than the last finalized block number, we should do an ancestor
+			// search to find a better common block.
+			if best_queued.saturating_sub(peer.common_number) > MAX_BLOCKS_TO_LOOK_BACKWARDS.into()
+				&& best_queued < peer.best_number && peer.common_number < last_finalized
```

Comment:
Why do we need the `peer.common_number < last_finalized` condition?

Reply:
If the common number is below the finalized number, the common number must be at least the last finalized.

Reply:
Got it. This is valid since we are guaranteeing that peer's best number is higher than our best (or queued). But what about chains with no finality where `last_finalized` will always be 0?

Reply:
You will run into other problems :see_no_evil: I added some todos for myself about further bugs, however I don't wanted to go deeper into this rabbit hole here :see_no_evil: